File: src/main/java/io/gittul/domain/like/LikeService.java
package io.gittul.domain.like;

import io.gittul.domain.comment.CommentRepository;
import io.gittul.domain.comment.entity.Comment;
import io.gittul.domain.like.entity.UserLikeComment;
import io.gittul.domain.like.entity.UserLikePost;
import io.gittul.domain.post.PostRepository;
import io.gittul.domain.post.entity.Post;
import io.gittul.domain.user.UserRepository;
import io.gittul.domain.user.entity.User;
import io.gittul.global.exception.CustomException;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class LikeService {

    private final PostRepository postRepository;
    private final CommentRepository commentRepository;
    private final UserRepository userRepository;

    public void likePost(User user, Long postId) {
        Post post = postRepository.getReferenceById(postId);
        UserLikePost like = UserLikePost.of(user, post);

        if (post.isLikedBy(user)) throw new CustomException(HttpStatus.CONFLICT, "이미 좋아요한 게시글입니다.");

        user.getDetails().getLikedPosts().add(like);
        userRepository.save(user);
    }

    public void unLikePost(User user, Long postId) {
        user.getDetails()
                .getLikedPosts()
                .removeIf(like -> like.getPost().getPostId().equals(postId));
        userRepository.save(user);
    }

    public void likeComment(User user, Long postId) {
        Comment comment = commentRepository.getReferenceById(postId);
        UserLikeComment like = UserLikeComment.of(user, comment);

        if (comment.isLikedBy(user)) throw new CustomException(HttpStatus.CONFLICT, "이미 좋아요한 댓글입니다.");

        user.getDetails().getLikedComments().add(like);
        userRepository.save(user);
    }

    public void unLikeComment(User user, Long postId) {
        user.getDetails()
                .getLikedComments()
                .removeIf(like -> like.getComment().getCommentId().equals(postId));
        userRepository.save(user);
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/like/LikeService.kt
package io.gittul.app.domain.like

import io.gittul.app.domain.comment.CommentRepository
import io.gittul.app.domain.thread.repository.ThreadRepository
import io.gittul.app.domain.user.UserRepository
import io.gittul.core.domain.like.entity.UserLikeComment
import io.gittul.core.domain.like.entity.userLikeThread
import io.gittul.core.domain.thread.entity.Thread
import io.gittul.core.domain.user.entity.User
import io.gittul.core.global.exception.CustomException
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class LikeService(
    private val threadRepository: ThreadRepository,
    private val commentRepository: CommentRepository,
    private val userRepository: UserRepository,
) {


    fun likeThread(user: User, threadId: Long) {
        val thread: Thread = threadRepository.getReferenceById(threadId)
        val like: userLikeThread = userLikeThread.of(user, thread)

        if (thread.isLikedBy(user)) throw CustomException(HttpStatus.CONFLICT, "이미 좋아요한 게시글입니다.")

        user.details.likedThreads.add(like)
        userRepository.save(user)
    }

    fun unLikeThread(user: User, threadId: Long) {
        user.details
            .likedThreads
            .removeIf { it.thread.threadId == threadId }
        userRepository.save(user)
    }

    fun likeComment(user: User, commentId: Long) {
        val comment = commentRepository.getReferenceById(commentId)
        val like: UserLikeComment = UserLikeComment.of(user, comment)

        if (comment.isLikedBy(user)) throw CustomException(
            HttpStatus.CONFLICT,
            "이미 좋아요한 댓글입니다."
        )

        user.details.likedComments.add(like)
        userRepository.save(user)
    }

    fun unLikeComment(user: User, commentId: Long) {
        user.details
            .likedComments
            .removeIf { it.comment.commentId == commentId }
        userRepository.save(user)
    }
}
